core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesUnassignedShardsAction.java

```diff
@@ -0,0 +1,179 @@
+
+    @Inject
+    public TransportIndicesUnassignedShardsAction(Settings settings, TransportService transportService, ClusterService clusterService, ThreadPool threadPool, ActionFilters actionFilters,
+                                                  TransportNodesListGatewayStartedShards listShardsInfo) {
+        super(settings, IndicesUnassignedShardsAction.NAME, transportService, clusterService, threadPool, actionFilters, IndicesUnassignedShardsRequest.class);
+        this.listShardsInfo = listShardsInfo;
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected IndicesUnassigedShardsResponse newResponse() {
+        return new IndicesUnassigedShardsResponse();
+    }
+
+    @Override
+    protected void masterOperation(IndicesUnassignedShardsRequest request, ClusterState state, ActionListener<IndicesUnassigedShardsResponse> listener) {
+        // collect all unassigned shards for the requested indices
+        DiscoveryNodes nodes = state.nodes();
+        MetaData metaData = state.metaData();
+        Set<String> requestedIndices = new HashSet<>();
+        requestedIndices.addAll(Arrays.asList(request.indices()));
+        List<ShardId> shardIdsToFetch = new ArrayList<>(state.routingNodes().unassigned().size());
+        for (MutableShardRouting shard : state.routingNodes().unassigned()) {
+            if (requestedIndices.size() == 0 || requestedIndices.contains(shard.shardId().index().getName())) {
+                shardIdsToFetch.add(shard.shardId());
+            }
+        }
+
+        // async fetch shard status from all the nodes
+        if (shardIdsToFetch.size() > 0) {
+            AsyncShardFetches asyncShardFetches = new AsyncShardFetches(nodes, metaData, shardIdsToFetch.size(), listener);
+            asyncShardFetches.fetch(shardIdsToFetch);
+        } else {
+            listener.onResponse(new IndicesUnassigedShardsResponse());
+        }
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(IndicesUnassignedShardsRequest request, ClusterState state) {
+        return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_READ, request.indices());
+    }
+
+    private class AsyncShardFetches {
+        private final DiscoveryNodes nodes;
+        private final MetaData metaData;
+        private final int expectedOps;
+        private final ActionListener<IndicesUnassigedShardsResponse> listener;
+        private final AtomicInteger opsCount;
+        private final Queue<InternalAsyncFetch.Response> fetchResponses;
+
+        AsyncShardFetches(DiscoveryNodes nodes, MetaData metaData, int expectedOps, ActionListener<IndicesUnassigedShardsResponse> listener) {
+            this.nodes = nodes;
+            this.metaData = metaData;
+            this.listener = listener;
+            this.expectedOps = expectedOps;
+            this.opsCount = new AtomicInteger(0);
```

Comment:
We have a class called `CountDown.java` for this - check it out it might make things simpler her